linuxkm/lkcapi_glue.c

```diff
@@ -0,0 +1,2537 @@
+        free(ctx->aes_encrypt);
+        ctx->aes_encrypt = NULL;
+    }
+    if (ctx->aes_decrypt) {
+        wc_AesFree(ctx->aes_decrypt);
+        free(ctx->aes_decrypt);
+        ctx->aes_decrypt = NULL;
+    }
+}
+
+static int km_AesSetKeyCommon(struct km_AesCtx * ctx, const u8 *in_key,
+                              unsigned int key_len, const char * name)
+{
+    int err;
+
+    err = wc_AesSetKey(ctx->aes_encrypt, in_key, key_len, NULL, AES_ENCRYPTION);
+
+    if (unlikely(err)) {
+        pr_err("error: km_AesSetKeyCommon %s failed: %d\n", name, err);
+        return err;
+    }
+
+    if (ctx->aes_decrypt) {
+        err = wc_AesSetKey(ctx->aes_decrypt, in_key, key_len, NULL, AES_DECRYPTION);
+
+        if (unlikely(err)) {
+            pr_err("error: km_AesSetKeyCommon %s failed: %d\n", name, err);
+            return err;
+        }
+    }
+
+    return 0;
+}
+
+#if defined(LINUXKM_LKCAPI_REGISTER_ALL) || \
+    defined(LINUXKM_LKCAPI_REGISTER_AESCBC) || \
+    defined(LINUXKM_LKCAPI_REGISTER_AESCFB)
+
+static void km_AesExit(struct crypto_skcipher *tfm)
+{
+    struct km_AesCtx * ctx = crypto_skcipher_ctx(tfm);
+    km_AesExitCommon(ctx);
+}
+
+#endif /* LINUXKM_LKCAPI_REGISTER_ALL ||
+        * LINUXKM_LKCAPI_REGISTER_AESCBC ||
+        * LINUXKM_LKCAPI_REGISTER_AESCFB
+        */
+
+#endif /* LINUXKM_LKCAPI_REGISTER_ALL || LINUXKM_LKCAPI_REGISTER_AESCBC ||
+        * LINUXKM_LKCAPI_REGISTER_AESCFB || LINUXKM_LKCAPI_REGISTER_AESGCM
+        */
+
+#if defined(HAVE_AES_CBC) && \
+    (defined(LINUXKM_LKCAPI_REGISTER_ALL) || defined(LINUXKM_LKCAPI_REGISTER_AESCBC))
+
+static int km_AesCbcInit(struct crypto_skcipher *tfm)
+{
+    struct km_AesCtx * ctx = crypto_skcipher_ctx(tfm);
+    return km_AesInitCommon(ctx, WOLFKM_AESCBC_DRIVER, 1);
+}
+
+static int km_AesCbcSetKey(struct crypto_skcipher *tfm, const u8 *in_key,
+                          unsigned int key_len)
+{
+    struct km_AesCtx * ctx = crypto_skcipher_ctx(tfm);
+    return km_AesSetKeyCommon(ctx, in_key, key_len, WOLFKM_AESCBC_DRIVER);
+}
+
+static int km_AesCbcEncrypt(struct skcipher_request *req)
+{
+    struct crypto_skcipher * tfm = NULL;
+    struct km_AesCtx *       ctx = NULL;
+    struct skcipher_walk     walk;
+    unsigned int             nbytes = 0;
+    int                      err = 0;
+
+    tfm = crypto_skcipher_reqtfm(req);
+    ctx = crypto_skcipher_ctx(tfm);
+
+    err = skcipher_walk_virt(&walk, req, false);
+
+    while ((nbytes = walk.nbytes)) {
+        err = wc_AesSetIV(ctx->aes_encrypt, walk.iv);
+
+        if (unlikely(err)) {
+            pr_err("wc_AesSetIV failed: %d\n", err);
+            return err;
+        }
+
+        err = wc_AesCbcEncrypt(ctx->aes_encrypt, walk.dst.virt.addr,
+                               walk.src.virt.addr, nbytes);
+
+        if (unlikely(err)) {
+            pr_err("wc_AesCbcEncrypt failed %d\n", err);
+            return err;
+        }
+
+        err = skcipher_walk_done(&walk, walk.nbytes - nbytes);
```

Comment:
When does walk.nbytes get updated?

Reply:
Also `skcipher_walk_next()` called from `skcipher_walk_done()` updates it to next walk stride amount if there are bytes remaining.

Reply:
fixed.